datastore/providers/postgres_datastore.py

```diff
@@ -0,0 +1,332 @@
+                    if "created_at" in list(metadata.keys()):
+                        if metadata["created_at"]:
+                            metadata["created_at"] = to_unix_timestamp(
+                                metadata["created_at"]
+                            )
+                    metadata = json.dumps(metadata)
+                    insert_statement = sql.SQL(
+                        "INSERT INTO {} (doc_id, chunk_id, text, embedding, metadata) VALUES ({}, {}, {}, {}, {})"
+                    ).format(
+                        sql.Identifier(POSTGRES_TABLENAME),
+                        sql.Literal(doc_id),
+                        sql.Literal(chunk.id),
+                        sql.Literal(chunk.text),
+                        sql.Literal(chunk.embedding),
+                        sql.Literal(metadata),
+                    )
+                    cur.execute(insert_statement)
+                    self.index_counter += 1
+
+        if POSTGRES_UPSERT_COMMAND == "COPY":
+            with cur.copy(
+                "COPY %s (doc_id, chunk_id, text, embedding, metadata) FROM STDIN"
+                % POSTGRES_TABLENAME
+            ) as copy:
+                for doc_id, chunk_list in chunks.items():
+                    doc_ids.append(doc_id)
+                    for chunk in chunk_list:
+                        text = sql.Literal(chunk.text).as_string(conn)
+                        embedding = (
+                            "[" + ",".join([str(v) for v in chunk.embedding]) + "]"
+                        )
+
+                        metadata = chunk.metadata.dict()
+                        if "created_at" in list(metadata.keys()):
+                            if metadata["created_at"]:
+                                metadata["created_at"] = to_unix_timestamp(
+                                    metadata["created_at"]
+                                )
+                        metadata = json.dumps(metadata)
+
+                        row = (doc_id, chunk.id, text, embedding, metadata)
+                        copy.write_row(row)
+                        self.index_counter += 1
+
+        cur.execute("SELECT COUNT(*) from %s" % POSTGRES_TABLENAME)
+        nrows = cur.fetchall()[0][0]
+
+        if self.index_exists:
+            if self.index_counter > POSTGRES_MIN_NEW_ROWS_FOR_REINDEX:
+                reindex_statement = "REINDEX INDEX CONCURRENTLY %s" % (
+                    POSTGRES_EMBEDDING_INDEX
+                )
+                cur.execute(reindex_statement)
+                print(
+                    "Reindex initiated when nrows = %d, index counter = %d"
+                    % (nrows, self.index_counter)
+                )
+                self.index_counter = 0
+
+        else:
+            if nrows > POSTGRES_MIN_ROWS_FOR_INDEX:
+                index_statement = (
+                    "CREATE INDEX CONCURRENTLY %s ON %s USING ivfflat (embedding vector_cosine_ops)"
```

Comment:
you'll want to pass `lists = count(*) / 1000`
```suggestion
                    "CREATE INDEX CONCURRENTLY %s ON %s USING ivfflat (embedding vector_cosine_ops)
                    WITH (lists = (SELECT count(*) / 1000 FROM docs))
                    "
```